helix-term/src/commands/lsp.rs

```diff
@@ -1367,14 +1368,39 @@ fn compute_inlay_hints_for_view(
                         None => continue,
                     };
 
-                let label = match hint.label {
+                let mut label = match hint.label {
                     lsp::InlayHintLabel::String(s) => s,
                     lsp::InlayHintLabel::LabelParts(parts) => parts
                         .into_iter()
                         .map(|p| p.value)
                         .collect::<Vec<_>>()
                         .join(""),
                 };
+                // Truncate the hint if too long
+                if let Some(limit) = inlay_hints_length_limit {
+                    // Limit on displayed width
+                    use helix_core::unicode::width::{UnicodeWidthChar, UnicodeWidthStr};
+
+                    let width = label.width();
+                    let limit = limit.get().into();
+                    if width > limit {
+                        // It's impossible for LSP to send a string of control chars, so 0 is fine
+                        let mut floor_boundary = 0;
+                        let mut acc = 0;
+                        for (i, c) in label.char_indices() {
```

Comment:
Instead of character indices we should look at grapheme indices using `helix_core::unicode::segmentation::UnicodeSegmentation`: if we truncate to a character boundary we could cut into a grapheme cluster, turning 🏴‍☠️ into 🏴 for example. (🏴‍☠️ is 🏴 with a zero-width joiner character and ☠️.)